lib/ansible/modules/net_tools/basics/get_url.py

```diff
@@ -529,37 +529,38 @@ def main():
             module.fail_json(msg='The checksum format is invalid', **result)
 
     if not dest_is_dir and os.path.exists(dest):
-        checksum_mismatch = False
+        # should start with checksum_match = False for more clarity
+        checksum_match = True
 
-        # If the download is not forced and there is a checksum, allow
-        # checksum match to skip the download.
-        if not force and checksum != '':
-            destination_checksum = module.digest_from_file(dest, algorithm)
+        if not force:
+            # If the download is not forced and there is a checksum, allow
+            # checksum match to skip the download.
+            if checksum != '':
+                destination_checksum = module.digest_from_file(dest, algorithm)
 
-            if checksum != destination_checksum:
-                checksum_mismatch = True
+                if checksum != destination_checksum:
+                    checksum_match = False
 
-        # Not forcing redownload, unless checksum does not match
-        if not force and checksum and not checksum_mismatch:
             # Not forcing redownload, unless checksum does not match
             # allow file attribute changes
-            module.params['path'] = dest
-            file_args = module.load_file_common_arguments(module.params)
-            file_args['path'] = dest
-            result['changed'] = module.set_fs_attributes_if_different(file_args, False)
-            if result['changed']:
-                module.exit_json(msg="file already exists but file attributes changed", **result)
-            module.exit_json(msg="file already exists", **result)
-
-        # If the file already exists, prepare the last modified time for the
-        # request.
-        mtime = os.path.getmtime(dest)
-        last_mod_time = datetime.datetime.utcfromtimestamp(mtime)
-
-        # If the checksum does not match we have to force the download
-        # because last_mod_time may be newer than on remote
-        if checksum_mismatch:
-            force = True
+            if (checksum and checksum_match) or not checksum:
```

Comment:
Condition 1:
```python
if (checksum and checksum_match) or not checksum:
```
Actually same as
Condition 2:
```python
if checksum_match:
```
because

|checksum|checksum_match|Condition 1|Condition 2|
|---|---|---|---|
|F|always T|T|T|
|T|T|T|T|
|T|F|F|F|

Reply:
Thanks for the 101 programming lesson @whs-dot-hk maybe @sivel and the team can review the specs and naming conventions on the get_url module together with @gtie to reduce confusion between contributors.

Reply:
I think it is not the naming

This piece of code is the logic for whether re-download or not

Reply:
@whs-dot-hk I think you made a request to work on this issue earlier in the week, if you have a solution you are welcome to take over.